packages/flutter/lib/src/widgets/shortcuts.dart

```diff
@@ -234,11 +371,211 @@ class ShortcutMapProperty extends DiagnosticsProperty<Map<LogicalKeySet, Intent>
        );
 
   @override
-  Map<LogicalKeySet, Intent> get value => super.value!;
+  Map<ShortcutActivator, Intent> get value => super.value!;
 
   @override
-  String valueToString({ TextTreeConfiguration? parentConfiguration }) {
-    return '{${value.keys.map<String>((LogicalKeySet keySet) => '{${keySet.debugDescribeKeys()}}: ${value[keySet]}').join(', ')}}';
+  String valueToString({TextTreeConfiguration? parentConfiguration}) {
+    return '{${value.keys.map<String>((ShortcutActivator keySet) => '{${keySet.debugDescribeKeys()}}: ${value[keySet]}').join(', ')}}';
+  }
+}
+
+/// A shortcut key combination of a single key and modifiers.
+///
+/// This [ShortcutActivator] implements the typical shortcuts such as:
+///
+///  * ArrowLeft
+///  * Shift + Delete
+///  * Control + Alt + Meta + Shift + A
+///
+/// More specifically, it create shortcut key combinations that are composed of a
+/// [trigger] key, and zero, some, or all of the four modifiers (control, shift,
+/// alt, meta). The shortcut is activated when the following conditions are met:
+///
+///  * The incoming event is a down event with the [trigger] key.
+///  * If [control] is true, then either or all control keys must be held.
+///    Otherwise, no control keys must be held.
+///  * The similar condition goes for [alt], [shift], and [meta].
+///
+/// This resembles the typical behavior of operation systems, and marks a major
+/// difference from [LogicalKeySet]:
+///
+///  * Additional non-modifier keys are allowed. For example,
+///    pressing key X while holding ControlLeft and key A *will* activate a
+///    `SingleActivator(LogicalKeyboardKey.keyX, control: true)`
+///  * The modifiers must not be the trigger key. For example, pressing
+///    ControlLeft while holding key X *will not* activate a
+///    `SingleActivator(LogicalKeyboardKey.keyX, control: true)`
+class SingleActivator with Diagnosticable implements ShortcutActivator {
+  /// Create an activator of a trigger key and modifiers.
+  ///
+  /// The `trigger` should be the non-modifier key that is pressed after all the
+  /// modifiers, such as [LogicalKeyboardKey.keyC] as in `Ctrl+C`. It must not be
+  /// a modifier key (sided or unsided).
```

Comment:
Modifier-only shortcuts are **not** a goal of `SingleActivator`.
We can provide another class in the future once we receive enough demands; or developers can define their own before that. Implementing `ShortcutActivator` is always possible.

Reply:
I don't think it's a good idea to limit the keys that can be used in a shortcut: people won't implement their own subclass, they'll just assume it can't be done. I think that if we're deprecating `LogicalKeySet`, we should add that other class (it could be the `KeySetActivator` that you're already planning) to give people a way to migrate without having to implement a new subclass.

Is the reason that you don't want modifier-only shortcuts to be a goal of `SingleActivator` because of the non-sidedness of the `control`, `meta`, etc. flags?

I can see that it might be confusing to say `SingleActivator(LogicalKeyboardKey.shiftLeft, shift: true)`, since now it's not clear whether you're talking about any shift or just `shiftLeft`, but you could just assert that the trigger isn't a `shift`, `shiftLeft` or `shiftRight` if the `shift` flag is true. Other than that, I don't see why the modifier keys should be excluded from being triggers.

Reply:
> you could just assert that the trigger isn't a shift, shiftLeft or shiftRight if the shift flag is true

If `shift` is not true, then it's false, which means that it checks that all shift keys should be released. The boolean here does not mean whether the key is checked, but the expected pressing state of the key during the check (and it would be confusing to change this semantic "only when the trigger is a modifier key"). It always checks.

> if we're deprecating LogicalKeySet, we should add that other class

We haven't deprecated `LogicalKeySet` yet. If we do, we can have another round of discussion the scope of the impact. This PR is meant to be non-breaking.

> people won't implement their own subclass, they'll just assume it can't be done

If there's a clear, sufficient demand and situation this can be used, we can always add a class for it. We can't add a new API just because we imagine the current API does not cover all possible combination. We never can.

Reply:
Ahh, OK, I didn't notice that it also enforces the key is not pressed. OK, then I guess it'll just have to be a separate class.